DatadogCore/Sources/Core/Upload/DataUploadWorker.swift

```diff
@@ -62,56 +63,67 @@ internal class DataUploadWorker: DataUploadWorkerType {
             guard let self = self else {
                 return
             }
-
             let context = contextProvider.read()
             let blockersForUpload = self.uploadConditions.blockersForUpload(with: context)
             let isSystemReady = blockersForUpload.isEmpty
-            let nextBatch = isSystemReady ? self.fileReader.readNextBatch() : nil
-            if let batch = nextBatch {
-                self.backgroundTaskCoordinator?.beginBackgroundTask()
-                DD.logger.debug("⏳ (\(self.featureName)) Uploading batch...")
-
-                do {
-                    // Upload batch
-                    let uploadStatus = try self.dataUploader.upload(
-                        events: batch.events,
-                        context: context
-                    )
-
-                    // Delete or keep batch depending on the upload status
-                    if uploadStatus.needsRetry {
-                        self.delay.increase()
-
-                        DD.logger.debug("   → (\(self.featureName)) not delivered, will be retransmitted: \(uploadStatus.userDebugDescription)")
-                    } else {
-                        self.fileReader.markBatchAsRead(batch, reason: .intakeCode(responseCode: uploadStatus.responseCode ?? -1)) // -1 is unexpected here
-                        self.delay.decrease()
-
-                        DD.logger.debug("   → (\(self.featureName)) accepted, won't be retransmitted: \(uploadStatus.userDebugDescription)")
+            let files = isSystemReady ? self.fileReader.readFiles(maxBatchesPerUpload) : nil
+            var allUploadsSucceeded = false
+            if let files = files, files.isEmpty == false {
+                for file in files {
```

Comment:
/blocker I don't think we should do a for loop within the work-item, otherwise we would be blocking one thread for the entire list of batches. Instead, we should create a `DispatchWorkItem` per batch upload combined in a `DispatchGroup`, this way we will balance the load between threads.

Reply:
> we would be blocking one thread for the entire list of batches

I agree 👍

> Instead, we should create a `DispatchWorkItem` per batch upload combined in a `DispatchGroup`, this way we will balance the load between threads

Distributing the load is one thing but the number of consumed threads is another. If I understand correctly @maxep , proposal is to use `DispatchGroup` and sync completion of multiple parallel `DispatchWorkItems`? That would explode the number of threads we use by orders of magnitude (think: 4 tracks, up to 100 batches).

Instead, IMO we should execute `DispatchWorkItems` sequentially on one queue, one after another (with no delay between them). This will link the number of threads to the number of tracks we use (so max 4).

Also, as we won't be drastically changing the routine of single `DispatchWorkItem`, this solution may not require changes to `Reader` interface (we will keep reading one batch per item like we do today). I'd also recommend to do a closer look at [this logic](https://github.com/DataDog/dd-sdk-ios/blob/05774d7c886efe98cf2cec662249fd96950cb790/DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift#L155-L165) after we implement the change. We hit problems in frequent directory enumerations in https://github.com/DataDog/dd-sdk-ios/pull/185. With the model proposed in this PR, we will be reading up to 400 batches sequentially in short delays, which requires extra measures.

Reply:
This is intentional. The idea is to read and upload all the selected batches without a delay synchronously.

Reply:
The idea with `DispatchGroup` is to notify the end of the upload to plan the next cycle, it was not necessarily about uploading in parallel. But indeed, we can only use `DispatchWorkItems` 👍 

@maciejburda, re. [your comment](https://github.com/DataDog/dd-sdk-ios/pull/1531#issuecomment-1805666977): Yes, the fact that we don't load all batches is best for memory, it's even better with the `autoreleaseFrequency: .workItem` property that we use for the upload queue.

My point here is about CPU, if you are running long blocking upload tasks, GDC will be forced to create new threads for other work items (form the host app), which is suboptiomal. See [Avoiding Excessive Thread Creation](https://developer.apple.com/documentation/dispatch/dispatchqueue#3136752).
Note that all the upload queues target the `.global(qos: .utility)` which is concurrent, meaning that you might be blocking N threads for long period, N being the number of registered Features.

If you use different work-items, you will give GDC the opportunity to dispatch the load on various threads, giving also the opportunity to run higher priority tasks in between each upload work-items.

Reply:
`DataUploadWorker` only performs single `DispatchWorkItem` at a time.